storage/innobase/xtrabackup/src/xbstream.cc

```diff
@@ -355,6 +367,8 @@ static file_entry_t *file_entry_new(extract_ctxt_t *ctxt, const char *path,
 
   if (ctxt->ds_decrypt_ctxt && ends_with(path, ".xbcrypt")) {
     file = ds_open(ctxt->ds_decrypt_ctxt, path, NULL);
+  } else if (ctxt->ds_decompress_ctxt && ends_with(path, ".qp")) {
```

Comment:
All those combinations should work. Anything ending with `.xbcrypt` goes to decrypt sink which, when `--decompress` is given, passes the file next to decompress sink. That'll check the file extension and if it is `.qp` it'll decompress but otherwise it'll just pass through to next sink without decompressing.